x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java

```diff
@@ -174,10 +364,64 @@ protected boolean initialRun() {
 
     @Override
     protected void onFinish(ActionListener<Void> listener) {
-        // reset the page size, so we do not memorize a low page size forever
-        pageSize = pivot.getInitialPageSize();
-        // reset the changed bucket to free memory
-        changedBuckets = null;
+        try {
+            // This indicates an early exit since no changes were found.
+            // So, don't treat this like a checkpoint being completed, as no work was done.
+            if (hasSourceChanged == false) {
+                listener.onResponse(null);
```

Comment:
Unless there is another `onFinish` that I missed, it seems we are no longer stopping if `shouldStopAtCheckpoint` is true here. That is a problem in the following scenario:

* Indexer kicked off to check for changes (state changed to Indexing from started)
* While the state is `Indexing` somebody called `_stop?wait_for_checkpoint=true`
* Stop sets `wait_for_checkpoint` to be true, and does not call stop (Indexer state is still INDEXING)
* OnFinish is called with no changes and the indexer changes to started
* Never stops even though `wait_for_checkpoint` is true.